### PR TITLE
Some TrueClass methods implemented in Ruby[Feature #17127]

### PR DESCRIPTION
**Pull request by S-H-GAMELINKS**
ref: [Feature #17127 Some TrueClass methods is faster if implemented in Ruby](https://bugs.ruby-lang.org/issues/17127)

### CODE REVIEW

File: benchmark/trueclass.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+benchmark:
+  to_s: |
+    true.to_s
+  inspect: |
+    true.inspect
+  or: |
+    true | false
+loop_count: 1000000
```
Comment (nobu) [inline benchmark/trueclass.yml:8]: No newline here.

File: object.c [hunk 2]
```diff
@@ -4672,12 +4634,7 @@ InitVM_Object(void)
     rb_deprecate_constant(rb_cObject, "Data");
 
     rb_cTrueClass = rb_define_class("TrueClass", rb_cObject);
-    rb_cTrueClass_to_s = rb_fstring_enc_lit("true", rb_usascii_encoding());
-    rb_gc_register_mark_object(rb_cTrueClass_to_s);
```
Comment (nobu) [inline object.c:4676]: A warning is emitted.
```
object.c:58:14: warning: unused variable 'rb_cTrueClass_to_s' [-Wunused-variable]
static VALUE rb_cTrueClass_to_s;
             ^
```

File: trueclass.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
+class TrueClass
+  #
+  # call-seq:
+  #   true.to_s   ->  "true"
+  #
+  # The string representation of <code>true</code> is "true".
+  #
+  def to_s
+    "true".freeze
```
Comment (nobu) [inline trueclass.rb:9]: This can be affected by the redefinition of `String#freeze`.
This should be `-"true"` or a frozen string literal.

**Review by nobu — COMMENTED** [review]
This makes `--dump=insns` weird.
```sh
$ ./ruby --dump=i trueclass.rb
== disasm: #<ISeq:<main>@trueclass.rb:1 (1,0)-(34,3)> (catch: FALSE)
0000 putspecialobject                       3                         (   3)[Li]
0002 putnil
0003 defineclass                            :TrueClass, <class:TrueClass>, 0
0007 leave

== disasm: #<ISeq:<class:TrueClass>@trueclass.rb:3 (3,0)-(34,3)> (catch: FALSE)
0000 definemethod                           :to_s, to_s               (  10)[LiCl]
0003 putself                                                          (  14)[Li]
0004 putobject                              :inspect
0006 putobject                              :to_s
0008 opt_send_without_block                 <calldata!mid:alias_method, argc:2, FCALL|ARGS_SIMPLE>
0010 pop
0011 definemethod                           :|, |                     (  31)[Li]
0014 putobject                              :|
0016 leave                                                            (  34)[En]

== disasm: #<ISeq:to_s@trueclass.rb:10 (10,2)-(12,5)> (catch: FALSE)
0000 putobject                              "true"                    (  11)[LiCa]
0002 leave                                                            (  12)[Re]

== disasm: #<ISeq:|@trueclass.rb:31 (31,2)-(33,5)> (catch: FALSE)
local table (size: 1, argc: 1 [opts: 0, rest: -1, post: 0, block: -1, kw: -1@-1, kwrest: -1])
[ 1] bool@0<Arg>
0000 putobject                              #<TrueClass:0x0000000000000014>(  32)[LiCa]
0002 leave                                                            (  33)[Re]
```

This is because builtin code is not installed when `--dump` option is given.

**Comment by S-H-GAMELINKS** [comment]
Corrected the review part.

File: trueclass.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  #
+  #  call-seq:
+  #     true & obj    -> true or false
+  #
+  #  And---Returns <code>false</code> if <i>obj</i> is
+  #  <code>nil</code> or <code>false</code>, <code>true</code> otherwise.
+  #
+  def &(obj)
+    Primitive.attr! 'inline'
+    Primitive.cexpr! 'RTEST(obj)?Qtrue:Qfalse;'
```
Comment (byroot) [inline trueclass.rb:31]: If the `cexpr!` even necessary here? Did you bench it against pure ruby?, e.g. `obj ? true : false`.

Comment (S-H-GAMELINKS) [reply trueclass.rb:31]: No, I'll try it.

Comment (S-H-GAMELINKS) [reply trueclass.rb:31]: pure ruby impl case as fast as using `cexpr!`
So, replace it in ruby code.

**Comment by S-H-GAMELINKS** [comment]
@nobu 
> This makes --dump=insns weird.

I'm not sure a bit, but what is strange about that?

**Comment by S-H-GAMELINKS** [comment]
`TrueClass#inspect` method defination added instead of use `alias_method`.

**Comment by S-H-GAMELINKS** [comment]
I thought be better re create this pull request, because it was better to put the files together.

**Pull request closed without merging**